search path demotion for llvm headers when building self-hosted with custom llvm-dependency

### Zig Version

0.10.0-dev.4110+dfcadd22b

### Steps to Reproduce

1. on archlinux...
2. install a version of llvm/clang/lld packages (so that headers are present in `/usr/include`) that are incompatible (eg. too old) to build zig against. This creates an interference with a custom llvm that you have built to build zig against.
3. build zig as per usual against custom llvm, expect building stage1 and stage2 to succeed, but stage3 to fail.

### Expected Behavior

self-hosted build to succeed

### Actual Behavior

self-hosted build fails because the wrong llvm header version is picked up and in this case zig's code is using a function that does not exist in `/usr/include`'s version.

```sh
error(compilation): clang failed with stderr: /home/mike/project/zig/work/main/src/stage1/codegen.cpp:4826:17: error: use of undeclared identifier 'LLVMPointerTypeInContext'
/home/mike/project/zig/work/main/src/stage1/codegen.cpp:5171:17: error: use of undeclared identifier 'LLVMPointerTypeInContext'

/home/mike/project/zig/work/main/src/stage1/codegen.cpp:1:1: error: unable to build C object: clang exited with code 1
error: zig...
error: The following command exited with error code 1:
/home/mike/project/zig/work/main/_build/zig2 build-exe --stack 33554432 /home/mike/project/zig/work/main/src/stage1.zig -lc -cflags -std=c++14 -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -D_GNU_SOURCE -fvisibility-inlines-hidden -fno-exceptions -fno-rtti -Werror=type-limits -Wno-missing-braces -Wno-comment -- /home/mike/project/zig/work/main/src/stage1/analyze.cpp /home/mike/project/zig/work/main/src/stage1/astgen.cpp /home/mike/project/zig/work/main/src/stage1/bigfloat.cpp /home/mike/project/zig/work/main/src/stage1/bigint.cpp /home/mike/project/zig/work/main/src/stage1/buffer.cpp /home/mike/project/zig/work/main/src/stage1/codegen.cpp /home/mike/project/zig/work/main/src/stage1/errmsg.cpp /home/mike/project/zig/work/main/src/stage1/error.cpp /home/mike/project/zig/work/main/src/stage1/heap.cpp /home/mike/project/zig/work/main/src/stage1/ir.cpp /home/mike/project/zig/work/main/src/stage1/ir_print.cpp /home/mike/project/zig/work/main/src/stage1/mem.cpp /home/mike/project/zig/work/main/src/stage1/os.cpp /home/mike/project/zig/work/main/src/stage1/parser.cpp /home/mike/project/zig/work/main/src/stage1/range_set.cpp /home/mike/project/zig/work/main/src/stage1/stage1.cpp /home/mike/project/zig/work/main/src/stage1/target.cpp /home/mike/project/zig/work/main/src/stage1/tokenizer.cpp /home/mike/project/zig/work/main/src/stage1/util.cpp /home/mike/project/zig/work/main/src/stage1/softfloat_ext.cpp -cflags -std=c99 -O3 -- /home/mike/project/zig/work/main/src/stage1/parse_f128.c /home/mike/project/zig/work/main/zig-cache/o/64d8a8cd48ea62ba1ffd9f5e9f0eaca8/libsoftfloat.a -lc++ /home/mike/project/zig/work/main/_build/zigcpp/libzigcpp.a /opt/llvm-linux-x86_64-15.0.0-release/lib/libclangFrontendTool.a /opt/llvm-linux-x86_64-15.0.0-release/lib/libclangCodeGen.a /opt/llvm-linux-x86_64-15.0.0-release/lib/libclangFrontend.a /opt/llvm-linux-x86_64-15.0.0-release/lib/libclangDriver.a /opt/llvm-linux-x86_64-15.0.0-release/lib/libclangSerialization.a /opt/llvm-linux-x86_64-15.0.0-release/lib/libclangSema.a /opt/llvm-linux-x86_64-15.0.0-release/lib/libclangStaticAnalyzerFrontend.a /opt/llvm-linux-x86_64-15.0.0-release/lib/libclangStaticAnalyzerCheckers.a /opt/llvm-linux-x86_64-15.0.0-release/lib/libclangStaticAnalyzerCore.a /opt/llvm-linux-x86_64-15.0.0-release/lib/libclangAnalysis.a /opt/llvm-linux-x86_64-15.0.0-release/lib/libclangASTMatchers.a /opt/llvm-linux-x86_64-15.0.0-release/lib/libclangAST.a /opt/llvm-linux-x86_64-15.0.0-release/lib/libclangParse.a /opt/llvm-linux-x86_64-15.0.0-release/lib/libclangSema.a /opt/llvm-linux-x86_64-15.0.0-release/lib/libclangBasic.a /opt/llvm-linux-x86_64-15.0.0-release/lib/libclangEdit.a /opt/llvm-linux-x86_64-15.0.0-release/lib/libclangLex.a /opt/llvm-linux-x86_64-15.0.0-release/lib/libclangARCMigrate.a /opt/llvm-linux-x86_64-15.0.0-release/lib/libclangRewriteFrontend.a /opt/llvm-linux-x86_64-15.0.0-release/lib/libclangRewrite.a /opt/llvm-linux-x86_64-15.0.0-release/lib/libclangCrossTU.a /opt/llvm-linux-x86_64-15.0.0-release/lib/libclangIndex.a /opt/llvm-linux-x86_64-15.0.0-release/lib/libclangToolingCore.a /opt/llvm-linux-x86_64-15.0.0-release/lib/libclangExtractAPI.a /opt/llvm-linux-x86_64-15.0.0-release/lib/libclangSupport.a /opt/llvm-linux-x86_64-15.0.0-release/lib/liblldMinGW.a /opt/llvm-linux-x86_64-15.0.0-release/lib/liblldELF.a /opt/llvm-linux-x86_64-15.0.0-release/lib/liblldCOFF.a /opt/llvm-linux-x86_64-15.0.0-release/lib/liblldWasm.a /opt/llvm-linux-x86_64-15.0.0-release/lib/liblldMachO.a /opt/llvm-linux-x86_64-15.0.0-release/lib/liblldCommon.a -lLLVMWindowsManifest -lLLVMWindowsDriver -lLLVMXRay -lLLVMLibDriver -lLLVMDlltoolDriver -lLLVMCoverage -lLLVMLineEditor -lLLVMXCoreDisassembler -lLLVMXCoreCodeGen -lLLVMXCoreDesc -lLLVMXCoreInfo -lLLVMX86TargetMCA -lLLVMX86Disassembler -lLLVMX86AsmParser -lLLVMX86CodeGen -lLLVMX86Desc -lLLVMX86Info -lLLVMWebAssemblyDisassembler -lLLVMWebAssemblyAsmParser -lLLVMWebAssemblyCodeGen -lLLVMWebAssemblyDesc -lLLVMWebAssemblyUtils -lLLVMWebAssemblyInfo -lLLVMVEDisassembler -lLLVMVEAsmParser -lLLVMVECodeGen -lLLVMVEDesc -lLLVMVEInfo -lLLVMSystemZDisassembler -lLLVMSystemZAsmParser -lLLVMSystemZCodeGen -lLLVMSystemZDesc -lLLVMSystemZInfo -lLLVMSparcDisassembler -lLLVMSparcAsmParser -lLLVMSparcCodeGen -lLLVMSparcDesc -lLLVMSparcInfo -lLLVMRISCVDisassembler -lLLVMRISCVAsmParser -lLLVMRISCVCodeGen -lLLVMRISCVDesc -lLLVMRISCVInfo -lLLVMPowerPCDisassembler -lLLVMPowerPCAsmParser -lLLVMPowerPCCodeGen -lLLVMPowerPCDesc -lLLVMPowerPCInfo -lLLVMNVPTXCodeGen -lLLVMNVPTXDesc -lLLVMNVPTXInfo -lLLVMMSP430Disassembler -lLLVMMSP430AsmParser -lLLVMMSP430CodeGen -lLLVMMSP430Desc -lLLVMMSP430Info -lLLVMMipsDisassembler -lLLVMMipsAsmParser -lLLVMMipsCodeGen -lLLVMMipsDesc -lLLVMMipsInfo -lLLVMLanaiDisassembler -lLLVMLanaiCodeGen -lLLVMLanaiAsmParser -lLLVMLanaiDesc -lLLVMLanaiInfo -lLLVMHexagonDisassembler -lLLVMHexagonCodeGen -lLLVMHexagonAsmParser -lLLVMHexagonDesc -lLLVMHexagonInfo -lLLVMBPFDisassembler -lLLVMBPFAsmParser -lLLVMBPFCodeGen -lLLVMBPFDesc -lLLVMBPFInfo -lLLVMAVRDisassembler -lLLVMAVRAsmParser -lLLVMAVRCodeGen -lLLVMAVRDesc -lLLVMAVRInfo -lLLVMARMDisassembler -lLLVMARMAsmParser -lLLVMARMCodeGen -lLLVMARMDesc -lLLVMARMUtils -lLLVMARMInfo -lLLVMAMDGPUTargetMCA -lLLVMAMDGPUDisassembler -lLLVMAMDGPUAsmParser -lLLVMAMDGPUCodeGen -lLLVMAMDGPUDesc -lLLVMAMDGPUUtils -lLLVMAMDGPUInfo -lLLVMAArch64Disassembler -lLLVMAArch64AsmParser -lLLVMAArch64CodeGen -lLLVMAArch64Desc -lLLVMAArch64Utils -lLLVMAArch64Info -lLLVMOrcJIT -lLLVMMCJIT -lLLVMJITLink -lLLVMInterpreter -lLLVMExecutionEngine -lLLVMRuntimeDyld -lLLVMOrcTargetProcess -lLLVMOrcShared -lLLVMDWP -lLLVMDebugInfoGSYM -lLLVMOption -lLLVMObjectYAML -lLLVMObjCopy -lLLVMMCA -lLLVMMCDisassembler -lLLVMLTO -lLLVMPasses -lLLVMCFGuard -lLLVMCoroutines -lLLVMObjCARCOpts -lLLVMipo -lLLVMVectorize -lLLVMLinker -lLLVMInstrumentation -lLLVMFrontendOpenMP -lLLVMFrontendOpenACC -lLLVMExtensions -lLLVMDWARFLinker -lLLVMGlobalISel -lLLVMMIRParser -lLLVMAsmPrinter -lLLVMSelectionDAG -lLLVMCodeGen -lLLVMIRReader -lLLVMAsmParser -lLLVMInterfaceStub -lLLVMFileCheck -lLLVMFuzzMutate -lLLVMTarget -lLLVMScalarOpts -lLLVMInstCombine -lLLVMAggressiveInstCombine -lLLVMTransformUtils -lLLVMBitWriter -lLLVMAnalysis -lLLVMProfileData -lLLVMSymbolize -lLLVMDebugInfoPDB -lLLVMDebugInfoMSF -lLLVMDebugInfoDWARF -lLLVMObject -lLLVMTextAPI -lLLVMMCParser -lLLVMMC -lLLVMDebugInfoCodeView -lLLVMBitReader -lLLVMFuzzerCLI -lLLVMCore -lLLVMRemarks -lLLVMBitstreamReader -lLLVMBinaryFormat -lLLVMTableGen -lLLVMSupport -lLLVMDemangle -lz /usr/lib/libzstd.so.1.5.2 /usr/lib/libstdc++.a -lunwind --cache-dir /home/mike/project/zig/work/main/zig-cache --global-cache-dir /home/mike/project/zig/work/main/zig-cache --name zig -target native-native -mcpu x86_64 --pkg-begin build_options /home/mike/project/zig/work/main/zig-cache/options/KT8v3sbeg53FNRkM2eKXbC07S8ZLh5UxAz4nTa54ORvuj1RFHAi8T-ztgwOol43P --pkg-end --pkg-begin compiler_rt /home/mike/project/zig/work/main/src/empty.zig --pkg-end -I /home/mike/project/zig/work/main/src -I /home/mike/project/zig/work/main/deps/SoftFloat-3e/source/include -I /home/mike/project/zig/work/main/deps/SoftFloat-3e-prebuilt -I /opt/llvm-linux-x86_64-15.0.0-release/include -I /opt/BOGUS_WTF -D ZIG_LINK_MODE=Static -L /opt/llvm-linux-x86_64-15.0.0-release/lib -isystem /opt/llvm-linux-x86_64-15.0.0-release/include -fno-build-id --zig-lib-dir /home/mike/project/zig/work/main/lib --enable-cache
```

To further drill down, after the failure, add `--verbose-cc` to the failed `zig2 build-exe` command to get the `zig2 clang` command where we add `-v` to show  header search path:
```sh
 "/home/mike/project/zig/work/main/_build/zig2" -cc1 -triple x86_64-unknown-linux-gnu -emit-obj -mrelax-all --mrelax-relocations -disable-free -clear-ast-before-backend -disable-llvm-verifier -discard-value-names -main-file-name codegen.cpp -mrelocation-model pic -pic-level 2 -fhalf-no-semantic-interposition -mframe-pointer=all -fmath-errno -ffp-contract=on -fno-rounding-math -mconstructor-aliases -funwind-tables=2 -target-cpu x86-64 -tune-cpu generic -mllvm -treat-scalable-fixed-error-as-warning -debug-info-kind=constructor -dwarf-version=4 -debugger-tuning=gdb -v -fcoverage-compilation-dir=/home/mike/project/zig/work/main -nostdsysteminc -nobuiltininc -resource-dir /home/mike/project/zig/work/main/lib/clang/15.0.0 -dependency-file /home/mike/project/zig/work/main/zig-cache/tmp/b72e986cfa10b073-codegen.o.d -MT /home/mike/project/zig/work/main/zig-cache/tmp/b72e986cfa10b073-codegen.o -sys-header-deps -MV -isystem /home/mike/project/zig/work/main/lib/libcxx/include -isystem /home/mike/project/zig/work/main/lib/libcxxabi/include -isystem /home/mike/project/zig/work/main/lib/libunwind/include -isystem /home/mike/project/zig/work/main/lib/include -isystem /usr/include -isystem /opt/llvm-linux-x86_64-15.0.0-release/include -isystem /usr/local/include -isystem /usr/include/x86_64-linux-gnu -isystem /usr/include -D _LIBCPP_DISABLE_VISIBILITY_ANNOTATIONS -D _LIBCXXABI_DISABLE_VISIBILITY_ANNOTATIONS -D _LIBCPP_HAS_NO_VENDOR_AVAILABILITY_ANNOTATIONS -D _LIBCPP_ABI_VERSION=1 -D _LIBCPP_ABI_NAMESPACE=__1 -D __GLIBC_MINOR__=36 -D _DEBUG -I /home/mike/project/zig/work/main/src -I /home/mike/project/zig/work/main/deps/SoftFloat-3e/source/include -I /home/mike/project/zig/work/main/deps/SoftFloat-3e-prebuilt -I /opt/llvm-linux-x86_64-15.0.0-release/include -I /opt/BOGUS_WTF -D ZIG_LINK_MODE=Static -D __STDC_CONSTANT_MACROS -D __STDC_FORMAT_MACROS -D __STDC_LIMIT_MACROS -D _GNU_SOURCE -O0 -Werror=type-limits -Wno-missing-braces -Wno-comment -std=c++14 -fdeprecated-macro -fdebug-compilation-dir=/home/mike/project/zig/work/main -ferror-limit 19 -fvisibility-inlines-hidden -fsanitize=alignment,array-bounds,bool,builtin,enum,float-cast-overflow,function,integer-divide-by-zero,nonnull-attribute,null,pointer-overflow,return,returns-nonnull-attribute,shift-base,shift-exponent,signed-integer-overflow,unreachable,vla-bound -fsanitize-trap=alignment,array-bounds,bool,builtin,enum,float-cast-overflow,function,integer-divide-by-zero,nonnull-attribute,null,pointer-overflow,return,returns-nonnull-attribute,shift-base,shift-exponent,signed-integer-overflow,unreachable,vla-bound -stack-protector 2 -stack-protector-buffer-size 4 -fno-rtti -fgnuc-version=4.2.1 -fno-caret-diagnostics -fcolor-diagnostics -fno-spell-checking -target-cpu x86-64 -target-feature -16bit-mode -target-feature -32bit-mode -target-feature -3dnow -target-feature -3dnowa -target-feature +64bit -target-feature -adx -target-feature -aes -target-feature -amx-bf16 -target-feature -amx-int8 -target-feature -amx-tile -target-feature -avx -target-feature -avx2 -target-feature -avx512bf16 -target-feature -avx512bitalg -target-feature -avx512bw -target-feature -avx512cd -target-feature -avx512dq -target-feature -avx512er -target-feature -avx512f -target-feature -avx512fp16 -target-feature -avx512ifma -target-feature -avx512pf -target-feature -avx512vbmi -target-feature -avx512vbmi2 -target-feature -avx512vl -target-feature -avx512vnni -target-feature -avx512vp2intersect -target-feature -avx512vpopcntdq -target-feature -avxvnni -target-feature -bmi -target-feature -bmi2 -target-feature -branchfusion -target-feature -cldemote -target-feature -clflushopt -target-feature -clwb -target-feature -clzero -target-feature +cmov -target-feature -crc32 -target-feature -cx16 -target-feature +cx8 -target-feature -enqcmd -target-feature -ermsb -target-feature -f16c -target-feature -false-deps-getmant -target-feature -false-deps-lzcnt-tzcnt -target-feature -false-deps-mulc -target-feature -false-deps-mullq -target-feature -false-deps-perm -target-feature -false-deps-popcnt -target-feature -false-deps-range -target-feature -fast-11bytenop -target-feature -fast-15bytenop -target-feature -fast-7bytenop -target-feature -fast-bextr -target-feature -fast-gather -target-feature -fast-hops -target-feature -fast-lzcnt -target-feature -fast-movbe -target-feature -fast-scalar-fsqrt -target-feature -fast-scalar-shift-masks -target-feature -fast-shld-rotate -target-feature -fast-variable-crosslane-shuffle -target-feature -fast-variable-perlane-shuffle -target-feature -fast-vector-fsqrt -target-feature -fast-vector-shift-masks -target-feature -fma -target-feature -fma4 -target-feature -fsgsbase -target-feature -fsrm -target-feature +fxsr -target-feature -gfni -target-feature -harden-sls-ijmp -target-feature -harden-sls-ret -target-feature -hreset -target-feature -idivl-to-divb -target-feature +idivq-to-divl -target-feature -invpcid -target-feature -kl -target-feature -lea-sp -target-feature -lea-uses-ag -target-feature -lvi-cfi -target-feature -lvi-load-hardening -target-feature -lwp -target-feature -lzcnt -target-feature +macrofusion -target-feature +mmx -target-feature -movbe -target-feature -movdir64b -target-feature -movdiri -target-feature -mwaitx -target-feature +nopl -target-feature -pad-short-functions -target-feature -pclmul -target-feature -pconfig -target-feature -pku -target-feature -popcnt -target-feature -prefer-128-bit -target-feature -prefer-256-bit -target-feature -prefer-mask-registers -target-feature -prefetchwt1 -target-feature -prfchw -target-feature -ptwrite -target-feature -rdpid -target-feature -rdpru -target-feature -rdrnd -target-feature -rdseed -target-feature -retpoline -target-feature -retpoline-external-thunk -target-feature -retpoline-indirect-branches -target-feature -retpoline-indirect-calls -target-feature -rtm -target-feature -sahf -target-feature -sbb-dep-breaking -target-feature -serialize -target-feature -seses -target-feature -sgx -target-feature -sha -target-feature -shstk -target-feature +slow-3ops-lea -target-feature +slow-incdec -target-feature -slow-lea -target-feature -slow-pmaddwd -target-feature -slow-pmulld -target-feature -slow-shld -target-feature -slow-two-mem-ops -target-feature -slow-unaligned-mem-16 -target-feature -slow-unaligned-mem-32 -target-feature -soft-float -target-feature +sse -target-feature +sse2 -target-feature -sse3 -target-feature -sse4.1 -target-feature -sse4.2 -target-feature -sse4a -target-feature -sse-unaligned-mem -target-feature -ssse3 -target-feature -tagged-globals -target-feature -tbm -target-feature -tsxldtrk -target-feature -uintr -target-feature -use-glm-div-sqrt-costs -target-feature -use-slm-arith-costs -target-feature -vaes -target-feature -vpclmulqdq -target-feature +vzeroupper -target-feature -waitpkg -target-feature -wbnoinvd -target-feature -widekl -target-feature +x87 -target-feature -xop -target-feature -xsave -target-feature -xsavec -target-feature -xsaveopt -target-feature -xsaves -faddrsig -D__GCC_HAVE_DWARF2_CFI_ASM=1 -o /home/mike/project/zig/work/main/zig-cache/tmp/b72e986cfa10b073-codegen.o -x c++ /home/mike/project/zig/work/main/src/stage1/codegen.cpp
clang -cc1 version 15.0.0 based upon LLVM 15.0.0 default target x86_64-unknown-linux-gnu
ignoring nonexistent directory "/opt/BOGUS_WTF"
ignoring nonexistent directory "/usr/include/x86_64-linux-gnu"
ignoring duplicate directory "/opt/llvm-linux-x86_64-15.0.0-release/include"
  as it is a non-system directory that duplicates a system directory
ignoring duplicate directory "/usr/include"
#include "..." search starts here:
#include <...> search starts here:
 /home/mike/project/zig/work/main/src
 /home/mike/project/zig/work/main/deps/SoftFloat-3e/source/include
 /home/mike/project/zig/work/main/deps/SoftFloat-3e-prebuilt
 /home/mike/project/zig/work/main/lib/libcxx/include
 /home/mike/project/zig/work/main/lib/libcxxabi/include
 /home/mike/project/zig/work/main/lib/libunwind/include
 /home/mike/project/zig/work/main/lib/include
 /usr/include
 /opt/llvm-linux-x86_64-15.0.0-release/include
 /usr/local/include
End of search list.
```

The issue cause is `/opt/llvm-linux-x86_64-15.0.0-release/include` (my custom llvm location) comes after `/usr/include`. The reasoning for this is a detail in clang command-line parsing. Here's what happens.

1. zig adds arg `-I /opt/llvm-linux-x86_64-15.0.0-release/include`. This is good.
2. zig adds arg (or already has added) `-isystem /usr/include`. This is good.
3. zig adds arg `-isystem /opt/llvm-linux-x86_64-15.0.0-release/include`. This bad as it interferes with previous `-I` option.  clang's beheaviour is to group `-I` searches together, in order as first seen on command-line. It also groups `-isystem` searches together as first seen on command line and in the end, the absolute search list first searches all `-I` dirs, then it searches all `-isystem` dirs. When clang gets a `-isystem` dir that interferes with `-I` it silently removes the dir from `-I` and adds it to `-isystem`, in effect demoting the search precedence of the dir.

I think we've had this behaviour for quite some time, and it explains a lot of "fuzzy" build issue reports with self-hosted people have been having over the last(?) year when a different version of llvm is already installed in a standard OS location such as `/usr/include`.

So what's the cause? This bit of code assumes search-prefix headers are `-isystem`. A simple fix may be to change this code to use `-I`. The other annoyance is we shouldn't have llvm headers listed twice on the command line. AFAICT, duplicate search dirs, as long as they use the same command switch, are harmless:
https://github.com/ziglang/zig/blob/2c3d87b168bf522f538e78325fdae71c320e6e20/lib/std/build.zig#L3122-L3125

note: the workaround to this issue is to temporarily remove the interfering llvm package from system

--- a/lib/std/build.zig
+++ b/lib/std/build.zig
@@ -3119,7 +3119,7 @@ pub const LibExeObjStep = struct {
             try zig_args.append(builder.pathJoin(&.{
                 search_prefix, "lib",
             }));
-            try zig_args.append("-isystem");
+            try zig_args.append("-I");
             try zig_args.append(builder.pathJoin(&.{
                 search_prefix, "include",
             }));

--- a/src/main.zig
+++ b/src/main.zig
@@ -415,7 +415,7 @@ const usage_build_generic =
     \\    plan9                   Plan 9 from Bell Labs object format
     \\    hex  (planned feature)  Intel IHEX
     \\    raw  (planned feature)  Dump machine code directly
-    \\  -dirafter [dir]           Add directory to AFTER include search path
+    \\  -idirafter [dir]          Add directory to AFTER include search path
     \\  -isystem  [dir]           Add directory to SYSTEM include search path
     \\  -I[dir]                   Add directory to include search path
     \\  -D[macro]=[value]         Define C [macro] to [value] (1 if [value] omitted)
@@ -844,10 +844,21 @@ fn buildOutputType(
                     defer it.i += 1;
                     return it.args[it.i];
                 }
+                fn nextOrFatal(it: *@This()) []const u8 {
+                    if (it.i >= it.args.len) {
+                        if (it.resp_file) |*resp| if (resp.next()) |sentinel| return std.mem.span(sentinel);
+                        fatal("expected parameter after {s}", .{it.args[it.i - 1]});
+                    }
+                    defer it.i += 1;
+                    return it.args[it.i];
+                }
             };
             var args_iter = Iterator{
                 .args = all_args[2..],
             };
+
+            var cssan = ClangSearchSanitizer.init(gpa, &clang_argv);
+            defer cssan.map.deinit();
 
             args_loop: while (args_iter.next()) |arg| {
                 if (mem.startsWith(u8, arg, "@")) {
@@ -894,9 +905,7 @@ fn buildOutputType(
                         cur_pkg = cur_pkg.parent orelse
                             fatal("encountered --pkg-end with no matching --pkg-begin", .{});
                     } else if (mem.eql(u8, arg, "--main-pkg-path")) {
-                        main_pkg_path = args_iter.next() orelse {
-                            fatal("expected parameter after {s}", .{arg});
-                        };
+                        main_pkg_path = args_iter.nextOrFatal();
                     } else if (mem.eql(u8, arg, "-cflags")) {
                         extra_cflags.shrinkRetainingCapacity(0);
                         while (true) {
@@ -914,67 +923,37 @@ fn buildOutputType(
                             fatal("expected [auto|on|off] after --color, found '{s}'", .{next_arg});
                         };
                     } else if (mem.eql(u8, arg, "--subsystem")) {
-                        const next_arg = args_iter.next() orelse {
-                            fatal("expected parameter after {s}", .{arg});
-                        };
-                        subsystem = try parseSubSystem(next_arg);
+                        subsystem = try parseSubSystem(args_iter.nextOrFatal());
                     } else if (mem.eql(u8, arg, "-O")) {
-                        optimize_mode_string = args_iter.next() orelse {
-                            fatal("expected parameter after {s}", .{arg});
-                        };
+                        optimize_mode_string = args_iter.nextOrFatal();
                     } else if (mem.eql(u8, arg, "--entry")) {
-                        entry = args_iter.next() orelse {
-                            fatal("expected parameter after {s}", .{arg});
-                        };
+                        entry = args_iter.nextOrFatal();
                     } else if (mem.eql(u8, arg, "--stack")) {
-                        const next_arg = args_iter.next() orelse {
-                            fatal("expected parameter after {s}", .{arg});
-                        };
+                        const next_arg = args_iter.nextOrFatal();
                         stack_size_override = std.fmt.parseUnsigned(u64, next_arg, 0) catch |err| {
                             fatal("unable to parse stack size '{s}': {s}", .{ next_arg, @errorName(err) });
                         };
                     } else if (mem.eql(u8, arg, "--image-base")) {
-                        const next_arg = args_iter.next() orelse {
-                            fatal("expected parameter after {s}", .{arg});
-                        };
+                        const next_arg = args_iter.nextOrFatal();
                         image_base_override = std.fmt.parseUnsigned(u64, next_arg, 0) catch |err| {
                             fatal("unable to parse image base override '{s}': {s}", .{ next_arg, @errorName(err) });
                         };
                     } else if (mem.eql(u8, arg, "--name")) {
-                        provided_name = args_iter.next() orelse {
-                            fatal("expected parameter after {s}", .{arg});
-                        };
+                        provided_name = args_iter.nextOrFatal();
                     } else if (mem.eql(u8, arg, "-rpath")) {
-                        try rpath_list.append(args_iter.next() orelse {
-                            fatal("expected parameter after {s}", .{arg});
-                        });
+                        try rpath_list.append(args_iter.nextOrFatal());
                     } else if (mem.eql(u8, arg, "--library-directory") or mem.eql(u8, arg, "-L")) {
-                        try lib_dirs.append(args_iter.next() orelse {
-                            fatal("expected parameter after {s}", .{arg});
-                        });
+                        try lib_dirs.append(args_iter.nextOrFatal());
                     } else if (mem.eql(u8, arg, "-F")) {
-                        try framework_dirs.append(args_iter.next() orelse {
-                            fatal("expected parameter after {s}", .{arg});
-                        });
+                        try framework_dirs.append(args_iter.nextOrFatal());
                     } else if (mem.eql(u8, arg, "-framework")) {
-                        const path = args_iter.next() orelse {
-                            fatal("expected parameter after {s}", .{arg});
-                        };
-                        try frameworks.put(gpa, path, .{});
+                        try frameworks.put(gpa, args_iter.nextOrFatal(), .{});
                     } else if (mem.eql(u8, arg, "-weak_framework")) {
-                        const path = args_iter.next() orelse {
-                            fatal("expected parameter after {s}", .{arg});
-                        };
-                        try frameworks.put(gpa, path, .{ .weak = true });
+                        try frameworks.put(gpa, args_iter.nextOrFatal(), .{ .weak = true });
                     } else if (mem.eql(u8, arg, "-needed_framework")) {
-                        const path = args_iter.next() orelse {
-                            fatal("expected parameter after {s}", .{arg});
-                        };
-                        try frameworks.put(gpa, path, .{ .needed = true });
+                        try frameworks.put(gpa, args_iter.nextOrFatal(), .{ .needed = true });
                     } else if (mem.eql(u8, arg, "-install_name")) {
-                        install_name = args_iter.next() orelse {
-                            fatal("expected parameter after {s}", .{arg});
-                        };
+                        install_name = args_iter.nextOrFatal();
                     } else if (mem.startsWith(u8, arg, "--compress-debug-sections=")) {
                         const param = arg["--compress-debug-sections=".len..];
                         linker_compress_debug_sections = std.meta.stringToEnum(link.CompressDebugSections, param) orelse {
@@ -983,9 +962,7 @@ fn buildOutputType(
                     } else if (mem.eql(u8, arg, "--compress-debug-sections")) {
                         linker_compress_debug_sections = link.CompressDebugSections.zlib;
                     } else if (mem.eql(u8, arg, "-pagezero_size")) {
-                        const next_arg = args_iter.next() orelse {
-                            fatal("expected parameter after {s}", .{arg});
-                        };
+                        const next_arg = args_iter.nextOrFatal();
                         pagezero_size = std.fmt.parseUnsigned(u64, eatIntPrefix(next_arg, 16), 16) catch |err| {
                             fatal("unable to parse pagezero size'{s}': {s}", .{ next_arg, @errorName(err) });
                         };
@@ -994,9 +971,7 @@ fn buildOutputType(
                     } else if (mem.eql(u8, arg, "-search_dylibs_first")) {
                         search_strategy = .dylibs_first;
                     } else if (mem.eql(u8, arg, "-headerpad")) {
-                        const next_arg = args_iter.next() orelse {
-                            fatal("expected parameter after {s}", .{arg});
-                        };
+                        const next_arg = args_iter.nextOrFatal();
                         headerpad_size = std.fmt.parseUnsigned(u32, eatIntPrefix(next_arg, 16), 16) catch |err| {
                             fatal("unable to parse headerpat size '{s}': {s}", .{ next_arg, @errorName(err) });
                         };
@@ -1007,65 +982,44 @@ fn buildOutputType(
                     } else if (mem.eql(u8, arg, "-dead_strip_dylibs")) {
                         dead_strip_dylibs = true;
                     } else if (mem.eql(u8, arg, "-T") or mem.eql(u8, arg, "--script")) {
-                        linker_script = args_iter.next() orelse {
-                            fatal("expected parameter after {s}", .{arg});
-                        };
+                        linker_script = args_iter.nextOrFatal();
                     } else if (mem.eql(u8, arg, "--version-script")) {
-                        version_script = args_iter.next() orelse {
-                            fatal("expected parameter after {s}", .{arg});
-                        };
+                        version_script = args_iter.nextOrFatal();
                     } else if (mem.eql(u8, arg, "--library") or mem.eql(u8, arg, "-l")) {
-                        const next_arg = args_iter.next() orelse {
-                            fatal("expected parameter after {s}", .{arg});
-                        };
                         // We don't know whether this library is part of libc or libc++ until
                         // we resolve the target, so we simply append to the list for now.
-                        try system_libs.put(next_arg, .{});
+                        try system_libs.put(args_iter.nextOrFatal(), .{});
                     } else if (mem.eql(u8, arg, "--needed-library") or
                         mem.eql(u8, arg, "-needed-l") or
                         mem.eql(u8, arg, "-needed_library"))
                     {
-                        const next_arg = args_iter.next() orelse {
-                            fatal("expected parameter after {s}", .{arg});
-                        };
+                        const next_arg = args_iter.nextOrFatal();
                         try system_libs.put(next_arg, .{ .needed = true });
                     } else if (mem.eql(u8, arg, "-weak_library") or mem.eql(u8, arg, "-weak-l")) {
-                        const next_arg = args_iter.next() orelse {
-                            fatal("expected parameter after {s}", .{arg});
-                        };
-                        try system_libs.put(next_arg, .{ .weak = true });
-                    } else if (mem.eql(u8, arg, "-D") or
-                        mem.eql(u8, arg, "-isystem") or
-                        mem.eql(u8, arg, "-I") or
-                        mem.eql(u8, arg, "-dirafter") or
-                        mem.eql(u8, arg, "-iwithsysroot") or
-                        mem.eql(u8, arg, "-iframework") or
-                        mem.eql(u8, arg, "-iframeworkwithsysroot"))
-                    {
+                        try system_libs.put(args_iter.nextOrFatal(), .{ .weak = true });
+                    } else if (mem.eql(u8, arg, "-D")) {
                         try clang_argv.append(arg);
-                        try clang_argv.append(args_iter.next() orelse {
-                            fatal("expected parameter after {s}", .{arg});
-                        });
+                        try clang_argv.append(args_iter.nextOrFatal());
+                    } else if (mem.eql(u8, arg, "-I")) {
+                        try cssan.addIncludePath(.I, arg, args_iter.nextOrFatal(), false);
+                    } else if (mem.eql(u8, arg, "-isystem") or mem.eql(u8, arg, "-iwithsysroot")) {
+                        try cssan.addIncludePath(.isystem, arg, args_iter.nextOrFatal(), false);
+                    } else if (mem.eql(u8, arg, "-idirafter")) {
+                        try cssan.addIncludePath(.idirafter, arg, args_iter.nextOrFatal(), false);
+                    } else if (mem.eql(u8, arg, "-iframework") or mem.eql(u8, arg, "-iframeworkwithsysroot")) {
+                        try cssan.addIncludePath(.iframework, arg, args_iter.nextOrFatal(), false);
                     } else if (mem.eql(u8, arg, "--version")) {
-                        const next_arg = args_iter.next() orelse {
-                            fatal("expected parameter after {s}", .{arg});
-                        };
+                        const next_arg = args_iter.nextOrFatal();
                         version = std.builtin.Version.parse(next_arg) catch |err| {
                             fatal("unable to parse --version '{s}': {s}", .{ next_arg, @errorName(err) });
                         };
                         have_version = true;
                     } else if (mem.eql(u8, arg, "-target")) {
-                        target_arch_os_abi = args_iter.next() orelse {
-                            fatal("expected parameter after {s}", .{arg});
-                        };
+                        target_arch_os_abi = args_iter.nextOrFatal();
                     } else if (mem.eql(u8, arg, "-mcpu")) {
-                        target_mcpu = args_iter.next() orelse {
-                            fatal("expected parameter after {s}", .{arg});
-                        };
+                        target_mcpu = args_iter.nextOrFatal();
                     } else if (mem.eql(u8, arg, "-mcmodel")) {
-                        machine_code_model = parseCodeModel(args_iter.next() orelse {
-                            fatal("expected parameter after {s}", .{arg});
-                        });
+                        machine_code_model = parseCodeModel(args_iter.nextOrFatal());
                     } else if (mem.startsWith(u8, arg, "-ofmt=")) {
                         target_ofmt = arg["-ofmt=".len..];
                     } else if (mem.startsWith(u8, arg, "-mcpu=")) {
@@ -1075,51 +1029,29 @@ fn buildOutputType(
                     } else if (mem.startsWith(u8, arg, "-O")) {
                         optimize_mode_string = arg["-O".len..];
                     } else if (mem.eql(u8, arg, "--dynamic-linker")) {
-                        target_dynamic_linker = args_iter.next() orelse {
-                            fatal("expected parameter after {s}", .{arg});
-                        };
+                        target_dynamic_linker = args_iter.nextOrFatal();
                     } else if (mem.eql(u8, arg, "--sysroot")) {
-                        sysroot = args_iter.next() orelse {
-                            fatal("expected parameter after {s}", .{arg});
-                        };
                         try clang_argv.append("-isysroot");
-                        try clang_argv.append(sysroot.?);
+                        try clang_argv.append(args_iter.nextOrFatal());
                     } else if (mem.eql(u8, arg, "--libc")) {
-                        libc_paths_file = args_iter.next() orelse {
-                            fatal("expected parameter after {s}", .{arg});
-                        };
+                        libc_paths_file = args_iter.nextOrFatal();
                     } else if (mem.eql(u8, arg, "--test-filter")) {
-                        test_filter = args_iter.next() orelse {
-                            fatal("expected parameter after {s}", .{arg});
-                        };
+                        test_filter = args_iter.nextOrFatal();
                     } else if (mem.eql(u8, arg, "--test-name-prefix")) {
-                        test_name_prefix = args_iter.next() orelse {
-                            fatal("expected parameter after {s}", .{arg});
-                        };
+                        test_name_prefix = args_iter.nextOrFatal();
                     } else if (mem.eql(u8, arg, "--test-cmd")) {
-                        try test_exec_args.append(args_iter.next() orelse {
-                            fatal("expected parameter after {s}", .{arg});
-                        });
+                        try test_exec_args.append(args_iter.nextOrFatal());
                     } else if (mem.eql(u8, arg, "--cache-dir")) {
-                        override_local_cache_dir = args_iter.next() orelse {
-                            fatal("expected parameter after {s}", .{arg});
-                        };
+                        override_local_cache_dir = args_iter.nextOrFatal();
                     } else if (mem.eql(u8, arg, "--global-cache-dir")) {
-                        override_global_cache_dir = args_iter.next() orelse {
-                            fatal("expected parameter after {s}", .{arg});
-                        };
+                        override_global_cache_dir = args_iter.nextOrFatal();
                     } else if (mem.eql(u8, arg, "--zig-lib-dir")) {
-                        override_lib_dir = args_iter.next() orelse {
-                            fatal("expected parameter after {s}", .{arg});
-                        };
+                        override_lib_dir = args_iter.nextOrFatal();
                     } else if (mem.eql(u8, arg, "--debug-log")) {
-                        const next_arg = args_iter.next() orelse {
-                            fatal("expected parameter after {s}", .{arg});
-                        };
                         if (!build_options.enable_logging) {
                             std.log.warn("Zig was compiled without logging enabled (-Dlog). --debug-log has no effect.", .{});
                         } else {
-                            try log_scopes.append(gpa, next_arg);
+                            try log_scopes.append(gpa, args_iter.nextOrFatal());
                         }
                     } else if (mem.eql(u8, arg, "--debug-link-snapshot")) {
                         if (!build_options.enable_link_snapshots) {
@@ -1128,9 +1060,7 @@ fn buildOutputType(
                             enable_link_snapshots = true;
                         }
                     } else if (mem.eql(u8, arg, "--entitlements")) {
-                        entitlements = args_iter.next() orelse {
-                            fatal("expected parameter after {s}", .{arg});
-                        };
+                        entitlements = args_iter.nextOrFatal();
                     } else if (mem.eql(u8, arg, "-fcompiler-rt")) {
                         want_compiler_rt = true;
                     } else if (mem.eql(u8, arg, "-fno-compiler-rt")) {
@@ -1319,9 +1249,7 @@ fn buildOutputType(
                     } else if (mem.eql(u8, arg, "-fno-allow-shlib-undefined")) {
                         linker_allow_shlib_undefined = false;
                     } else if (mem.eql(u8, arg, "-z")) {
-                        const z_arg = args_iter.next() orelse {
-                            fatal("expected parameter after {s}", .{arg});
-                        };
+                        const z_arg = args_iter.nextOrFatal();
                         if (mem.eql(u8, z_arg, "nodelete")) {
                             linker_z_nodelete = true;
                         } else if (mem.eql(u8, z_arg, "notext")) {
@@ -1391,10 +1319,10 @@ fn buildOutputType(
                         try system_libs.put(arg["-needed-l".len..], .{ .needed = true });
                     } else if (mem.startsWith(u8, arg, "-weak-l")) {
                         try system_libs.put(arg["-weak-l".len..], .{ .weak = true });
-                    } else if (mem.startsWith(u8, arg, "-D") or
-                        mem.startsWith(u8, arg, "-I"))
-                    {
+                    } else if (mem.startsWith(u8, arg, "-D")) {
                         try clang_argv.append(arg);
+                    } else if (mem.startsWith(u8, arg, "-I")) {
+                        try cssan.addIncludePath(.I, arg, arg[2..], true);
                     } else if (mem.startsWith(u8, arg, "-mexec-model=")) {
                         wasi_exec_model = std.meta.stringToEnum(std.builtin.WasiExecModel, arg["-mexec-model=".len..]) orelse {
                             fatal("expected [command|reactor] for -mexec-mode=[value], found '{s}'", .{arg["-mexec-model=".len..]});
@@ -5485,3 +5413,69 @@ fn parseSubSystem(next_arg: []const u8) !std.Target.SubSystem {
         });
     }
 }
+
+/// Model a header searchlist as a group.
+/// Silently ignore superfluous search dirs.
+/// Warn when a dir is added to multiple searchlists.
+const ClangSearchSanitizer = struct {
+    argv: *std.ArrayList([]const u8),
+    map: std.StringHashMap(Membership),
+
+    fn init(gpa: Allocator, argv: *std.ArrayList([]const u8)) @This() {
+        return .{
+            .argv = argv,
+            .map = std.StringHashMap(Membership).init(gpa),
+        };
+    }
+
+    fn addIncludePath(self: *@This(), group: Group, arg: []const u8, dir: []const u8, joined: bool) !void {
+        const gopr = try self.map.getOrPut(dir);
+        const m = gopr.value_ptr;
+        if (!gopr.found_existing) {
+            // init empty membership
+            m.* = .{};
+        }
+        const wtxt = "add '{s}' to header searchlist '-{s}' conflicts with '-{s}'";
+        switch (group) {
+            .I => {
+                if (m.I) return;
+                m.I = true;
+                if (m.isystem) std.log.warn(wtxt, .{ dir, "I", "isystem" });
+                if (m.idirafter) std.log.warn(wtxt, .{ dir, "I", "idirafter" });
+                if (m.iframework) std.log.warn(wtxt, .{ dir, "I", "iframework" });
+            },
+            .isystem => {
+                if (m.isystem) return;
+                m.isystem = true;
+                if (m.I) std.log.warn(wtxt, .{ dir, "isystem", "I" });
+                if (m.idirafter) std.log.warn(wtxt, .{ dir, "isystem", "idirafter" });
+                if (m.iframework) std.log.warn(wtxt, .{ dir, "isystem", "iframework" });
+            },
+            .idirafter => {
+                if (m.idirafter) return;
+                m.idirafter = true;
+                if (m.I) std.log.warn(wtxt, .{ dir, "idirafter", "I" });
+                if (m.isystem) std.log.warn(wtxt, .{ dir, "idirafter", "isystem" });
+                if (m.iframework) std.log.warn(wtxt, .{ dir, "idirafter", "iframework" });
+            },
+            .iframework => {
+                if (m.iframework) return;
+                m.iframework = true;
+                if (m.I) std.log.warn(wtxt, .{ dir, "iframework", "I" });
+                if (m.isystem) std.log.warn(wtxt, .{ dir, "iframework", "isystem" });
+                if (m.idirafter) std.log.warn(wtxt, .{ dir, "iframework", "idirafter" });
+            },
+        }
+        try self.argv.append(arg);
+        if (!joined) try self.argv.append(dir);
+    }
+
+    const Group = enum { I, isystem, idirafter, iframework };
+
+    const Membership = packed struct {
+        I: bool = false,
+        isystem: bool = false,
+        idirafter: bool = false,
+        iframework: bool = false,
+    };
+};

